idx/memory/write_queue.go

```diff
@@ -0,0 +1,106 @@
+package memory
+
+import (
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/schema"
+)
+
+type WriteQueue struct {
+	sync.RWMutex
+	archives    map[schema.MKey]*idx.Archive
+	shutdown    chan struct{}
+	done        chan struct{}
+	maxBuffered int
+	maxDelay    time.Duration
+	flushed     chan struct{}
+
+	idx *UnpartitionedMemoryIdx
+}
+
+// NewWriteQueue creates a new writeQueue that will add archives to the passed UnpartitionedMemoryIdx
+// in batches
+func NewWriteQueue(index *UnpartitionedMemoryIdx, maxDelay time.Duration, maxBuffered int) *WriteQueue {
+	wq := &WriteQueue{
+		archives:    make(map[schema.MKey]*idx.Archive),
+		shutdown:    make(chan struct{}),
+		done:        make(chan struct{}),
+		maxBuffered: maxBuffered,
+		maxDelay:    maxDelay,
+		flushed:     make(chan struct{}, 1),
+		idx:         index,
+	}
+	go wq.loop()
+	return wq
+}
+
+func (wq *WriteQueue) Stop() {
+	close(wq.shutdown)
+	<-wq.done
+}
+
+func (wq *WriteQueue) Queue(archive *idx.Archive) {
+	wq.Lock()
+	wq.archives[archive.Id] = archive
```

Comment:
Is it possible that in a given time period that this will overwrite a previous map entry before the queue is flushed / items added to the index, and would that cause any problems? I know there is a check already for `metricPoints` which reaches back into the queue to see if the item is there which is called from `Update`, but what about in `AddOrUpdate`?

Reply:
I deliberately used a map instead of a slice to allow new archives to replace existing ones.